### PR TITLE
Fix some ndimage.interpolation endianness bugs

### PR DESCRIPTION
**Pull request by cdouglass**
* 1d affine transform with output parameter given as an ndarray with non-native endianness (#7388)
* `affine_transform`, `geometric_transform`, and `map_coordinates` with output parameter given as a dtype with non-native endianness

Applied a `_fix_endianness` decorator to these three functions, replacing the existing `_geometric_transform` wrapper that fixed endianness when output parameter was given as an ndarray to `geometric_transform`, `map_coordinates`, and >1d case of `affine_transform`.

https://github.com/scipy/scipy/issues/7388#issuecomment-306246625 indicates this sort of issue is more widespread but I am undecided whether to cover more cases in this PR or leave it for later.

### CODE REVIEW

**Comment by cdouglass** [comment]
For reviewer(s): I modeled the tests on those in #4150 but please double check I'm testing the right things.

**Comment by cdouglass** [comment]
I've decided not to cover more cases in this PR, so this is ready for review.

**Review by jaimefrio — CHANGES_REQUESTED** [review]
Thanks for working on this!

The decorator looks great, and we should be able to reuse this code elsewhere in this module. I have one request about preserving the original function signature there.

My complaints about the test are to be considered more like suggestions: if you find it in you to improve them, please do. But given the general state of all those tests, I think it's also OK to leave thing as they are for now.

File: scipy/ndimage/tests/test_ndimage.py
Comment (cdouglass) [inline scipy/ndimage/tests/test_ndimage.py]: Done!

File: scipy/ndimage/tests/test_ndimage.py
Comment (cdouglass) [inline scipy/ndimage/tests/test_ndimage.py]: Done.

File: scipy/ndimage/interpolation.py
Comment (cdouglass) [inline scipy/ndimage/interpolation.py]: Done.

File: scipy/ndimage/interpolation.py
Comment (cdouglass) [inline scipy/ndimage/interpolation.py]: Thanks, I didn't know about that. Done.

**Review by jaimefrio — APPROVED** [review]

**Comment by jaimefrio** [comment]
The code now LGTM, thanks again for taking the time to do this!

There are some minor nitpicks regarding github etiquette that may need resolving before hitting the merge button, but I'd appreciate if someone more familiar with the rules could weigh in:

 - I think the commits should be squashed (see [here](https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#rewriting-commit-history)) into a single one.
 - I think the PR description should also follow the commit naming rules, i.e. begin with `BUG`
 - if you include `fixes #7388` in the PR's description, github will close the corresponding bug automatically.

I'll give it a day for someone else to confirm the need to do any of this, and if no one complains I'll merge it tomorrow night.

**Comment by rgommers** [comment]
> I think the commits should be squashed (see here) into a single one.

The 5 commits here don't look like way too many to me - it would have been nice to squash commit 6cd5fd5c (that's a couple of style changes only), but the other ones are all sensible units as individual commits. Squashing would be fine but is optional I'd say.

> I think the PR description should also follow the commit naming rules, i.e. begin with BUG

I usually do that in the merge commit if it bothers me, or fix up the PR title myself if it's too nondescriptive. Normally wouldn't ask a new contributor to do that.

> if you include fixes #7388 in the PR's description, github will close the corresponding bug automatically.

The first commit message says ``Resolves #7388`` which will close the issue, so this is fine.

Hit the green button I'd say Jaime.

**Comment by cdouglass** [comment]
I've rebased into two commits. Thanks for the feedback, everyone!

**Comment by jaimefrio** [comment]
In it goes. Thanks Cathy!

**Comment by lesteve** [comment]
Great stuff, thanks a lot !

**Pull request merged**
